### PR TITLE
WIP: Add distance metrics

### LINKED ISSUE #39: support Jaccard and angular similarity calculations
these are both calculated in sourmash, albeit for FracMinHash sketches.

ref for angular similarity: [wikipedia](https://en.wikipedia.org/wiki/Cosine_similarity#Angular_distance_and_similarity)

abundance projection has also been useful over in sourmash, in which the abundances from one set of k-mers are projected onto the (flat) k-mers from another; the specific use case in sourmash has been to support reference genome abundance estimation in metagenomes. we might do something similar here.

### PR DESCRIPTION
**Pull request by Adamtaranto**
This PR will add simple distance metrics for comparing count tables. Starting with Jaccard and Cosine Similarity.

Closes #39 

This initial commit adds `cosine()` method to calculate cosine similarity between two `KmerCountTables`.
Have attempted to use Rayon to make hash lookup multithreaded.

@ctb can you check the logic on this + look over the tests?

I've benchmarked against the scipy cosine function (allowing for a 0.001% margin - not sure where the diff is coming from)

In the test `test_cosine_similarity_identical_tables()` my function is giving 0.999999999 instead of 1. No idea why this is happening.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -440,6 +440,68 @@ impl KmerCountTable {
         self.counts.insert(hashval, count);
         Ok(())
     }
+
+    /// Calculates the Jaccard Similarity Coefficient between two KmerCountTable objects.
+    /// # Returns
+    /// The Jaccard Similarity Coefficient between the two tables as a float value between 0 and 1.
+    pub fn jaccard(&self, other: &KmerCountTable) -> f64 {
```
Comment (ctb) [inline src/lib.rs:447]: 👍

File: src/lib.rs [hunk 2]
```diff
@@ -440,6 +440,68 @@ impl KmerCountTable {
...
+            return 0.0;
+        }
+
+        // Calculate the dot product in parallel.
+        let dot_product: u64 = self
+            .counts
+            .par_iter()
+            .filter_map(|(&hash, &count1)| {
+                // Only include in the dot product if both tables have the k-mer.
+                other.counts.get(&hash).map(|&count2| count1 * count2)
```
Comment (ctb) [inline src/lib.rs:478]: I don't immediately understand the error handling here - `get` returns a `None` if `&hash` is not present, but does `.map` operate on a `None`??

**Comment by ctb** [comment]
see https://github.com/oxli-bio/oxli/pull/48 - I added a bit to the tests. Overall this looks great, thank you!!

File: src/lib.rs (hunk 2 above)
Comment (Adamtaranto) [reply src/lib.rs:478]: I don't think there should be any errors to handle, `get()` will return `None` if the query hash is not present in the _other_ table. I think `map()` should skip anything that returns `None` and not attempt to multiply count1 * count2.

Seeing as we sum all the results at the end the missing zeros shouldn't make a difference.

**Pull request merged**